controller/apps.go

```diff
@@ -0,0 +1,465 @@
+	// get the user definition (for cluster URL)
+	resp, err := authClient.ShowUser(ctx, authservice.ShowUserPath(), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, fmt.Errorf("Failed to GET user due to status code %d", status)
+	}
+
+	var respType authservice.User
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+func getTokenData(authClient authservice.Client, ctx context.Context, forService string) (*authservice.TokenData, error) {
+
+	resp, err := authClient.RetrieveToken(ctx, authservice.RetrieveTokenPath(), forService, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, errors.New("Failed to GET user due to status code " + string(status))
+	}
+
+	var respType authservice.TokenData
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+// getKubeClient createa kube client for the appropriate cluster assigned to the current user.
+// many different errors are possible, so controllers should call getAndCheckKubeClient() instead
+func (c *AppsController) getKubeClient(ctx context.Context) (*kubernetes.KubeClient, error) {
+
+	// create Auth API client
+	authClient, err := auth.CreateClient(ctx, c.Config)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	authUser, err := getUser(*authClient, ctx)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	if authUser == nil || authUser.Data.Attributes.Cluster == nil {
+		log.Error(ctx, nil, "error getting user from Auth server:"+tostring(authUser))
+		return nil, fmt.Errorf("error getting user from Auth Server: %s", tostring(authUser))
+	}
+
+	// get the openshift/kubernetes auth info for the cluster OpenShift API
+	osauth, err := getTokenData(*authClient, ctx, *authUser.Data.Attributes.Cluster)
+	if err != nil {
+		log.Error(ctx, nil, "error getting openshift credentials:"+tostring(err))
+		return nil, err
+	}
+
+	kubeURL := *authUser.Data.Attributes.Cluster
+	kubeToken := *osauth.AccessToken
+
+	kubeNamespaceName, err := getNamespaceName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the cluster API client
+	kc, err := kubernetes.NewKubeClient(kubeURL, kubeToken, *kubeNamespaceName)
+	if err != nil {
+		return nil, err
+	}
+	return kc, nil
+}
+
+// getAndCheckKubeClient converts all errors fromgetKubeClient() to Error 401
+// this is for convenience and ensuring consistency
+func (c *AppsController) getAndCheckKubeClient(ctx context.Context) (*kubernetes.KubeClient, error) {
+
+	kc, err := c.getKubeClient(ctx)
+	if err != nil {
```

Comment:
this `getAndCheckKubeClient` seems unnecessary: ` c.getKubeClient(ctx)` already returns an `error`, so you could do the conversion to `NewUnauthorizedError(...)` from the caller (which also checks for an error anyways)